lib/rbac/filterer.rb

```diff
@@ -254,10 +254,22 @@ def search(options = {})
 
       scope = include_references(scope, klass, include_for_find, exp_includes, skip_references)
       scope = scope.limit(limit).offset(offset) if attrs[:apply_limit_in_sql]
+
+      if inline_view?(options, scope)
+        inner_scope = scope.except(:select, :includes, :references)
+        scope.includes_values.each { |hash| inner_scope = add_joins(klass, inner_scope, hash) }
+        scope = scope.from(Arel.sql("(#{inner_scope.to_sql})").as(scope.table_name))
+                     .except(:offset, :limit, :order, :where)
+
+        # the auth_count needs to come from the inner query (the query with the limit)
+        if !options[:skip_counts] && (attrs[:apply_limit_in_sql] && limit)
+          auth_count = inner_scope.except(:offset, :limit, :order).count(:all)
```

Comment:
Is it possible that a filter on the outer_scope will further shrink the record count?  That is, the inner_scope limit 20 returns 20 records, but then something outside it (maybe rbac, but even just a "name" filter) filters away 10 of those.  So, not would the auth_count be wrong, but then so is what the caller actually requested.  Can you add a spec for that case, maybe?

Reply:
the `where()` clause is kept in `inner_scope` and removed from `scope`
The rbac portion has already been applied.

`auth_count` is only calculated here if we `apply_limit_in_sql`. So if we are filtering out records outside of sql, then this count will not be calculated.

also note the if clause above matches the `? :` condition below.
So this is just executing the identical sql `count` a little bit earlier.

But please keep me honest here

Reply:
Ok, one scenario: We have `apply_limit_in_sql` and also have a name filter, possibly due to the advanced search MiqExpression is both sql and non-sql friendly.

I don't think it is possible, but if it were...

Before this code change, it would run the sql count the same way (2 lines earlier) and it will filter the names a little below, subtracting from the total.

So I think it is bug for bug compatible.

Reply:
I need examples in tests to better understand this.  I believe Keenan is working on that to verify it.

Reply:
ok verified.

rbac filtered a few of the records.
sql filtered a few more
ruby filtered a few more

it got the right records and counts out.

Reply:
The answer:

So, if you have sql and ruby clauses in an miq_expression, it is marked `supports_sql = false`, even though a little sql does indeed come out and is applied to the where clause.

`apply_limit_in_sql = false` - since we really do need to bring back all records before we filter and then limit them.

This PR ends up using an inline view for virtual attributes, and the counts / limit in ruby works fine.
Since there is some code in the where clause, the number of times the subquery is called is reduced.

The test shows the numbers worked out and it properly filtered the data